app/models/ability.rb

```diff
@@ -60,6 +60,14 @@ def initialize(user)
         can? :teach, sub.course
       end
 
+      can :download, Submission do |sub|
+        !sub.course.hide_submission_result? && can?(:read, sub)
+      end
+
+      can :read_result, Submission do |sub|
+        !sub.course.hide_submission_result? || can?(:teach, sub.course)
```

Comment:
using brackets with can? clause is different than elsewhere in file. not sure which is best, but either should use this, or (can? :read, sub) -style in whole file i think